### PR TITLE
Add a patttern for running data migration

### PR DESCRIPTION
**Pull request by kxhitiz**
Followup keynotes: 
- Down the road, once we have this process established, we can have a generator script to automatically generate the scaffold files for the data migration.
- Over time, we will have a huge list of migration scripts, and we will need to have some sort of scheduled cleanup process to remove legacy migration scripts

### CODE REVIEW

File: _patterns/running-data-migration.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+- **Non Trivial Script Execution:**
+  - If you have script that needs to be run, potentially multiple times:
+    - Send it to rake task under the one-off migration section (`/lib/tasks/one_off/migrations.rake`).
+- **Quick Small Changes in Database Records:**
+  - For quick, small changes in database records:
+    - What qualifies as small change?
+      - attribute change to a single record
+    - Access the console and make the necessary changes
+    - Maintain an audit trail to understand why and what changes were made (https://github.com/basecamp/audits1984)
```
Comment (kxhitiz) [inline _patterns/running-data-migration.md:24]: This point is not relevant at the moment until [this](https://biggerpockets.atlassian.net/browse/BIG-12537) ticket is deployed

File: _patterns/running-data-migration.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+## How to create & run my script
+To create and run your script, follow these steps:
+
+- **Write Your Script as a Rake Task**
+  - Create a class within migration directory script (eg. `/lib/migrations/your_script.rb`)
```
Comment (kxhitiz) [inline _patterns/running-data-migration.md:30]:
```suggestion
  - Create a class within migration directory script (eg. `/lib/migrations/do_x.rb`)
```

**Review by karlentwistle — APPROVED** [review]

**Review by RenatoCoelho-BiggerPockets — APPROVED** [review]

**Review by bhanubhakta — APPROVED** [review]

**Review by mgallagher0 — APPROVED** [review]
Used this as a template the other week for running a one-off migration and it was very helpful. Thanks Kshitiz 👍

**Review by cernivec — APPROVED** [review]

File: _patterns/best_practices_on_writing_a_script.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  - Keep a record of the progress made during the execution of the script
+- **Record Before and After States:**
+  - Document the states of the data before and after script execution
+- **Implement Dry Mode:**
+  - Include a dry mode in your script to allow running it without making actual changes to database records, 
+    providing a preview of the run. It helps us realize the extent of the actual change the script will make.
+  - eg.
+    ```ruby
+    class DoX
+      def initialize(dry: true)
```
Comment (kxhitiz) [inline _patterns/best_practices_on_writing_a_script.md:18]:
```suggestion
      def initialize(dry_mode: true)
```

File: _patterns/running-data-migration.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    - What qualifies as small change?
+      - attribute change to a single record
+    - Access the console and make the necessary changes
+    - Maintain an audit trail to understand why and what changes were made (https://github.com/basecamp/audits1984)
+
+## How to create & run my script
+To create and run your script, follow these steps:
+
+- **Write Your Script as a Rake Task**
+  - Create a class within migration directory script (eg. `/lib/migrations/do_x.rb`)
```
[the remark this answers is not in the document]
Comment (kxhitiz) [reply _patterns/running-data-migration.md:30]: We have our older scripts within the `lib/migrations` directory, hence I went with that data structure.

File: _patterns/running-data-migration.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  - Add your task to list of one_off tasks in the file (`/lib/tasks/one_off/migrations.rake`)
+  - eg.
+    ```ruby
+    namespace :one_off do
+      namespace :migrations do
+        ...
+
+        desc "This tasks does x"
+        task do_x: :environment do
+          Migrations::DoX.new.call
```
[the remark this answers is not in the document]
Comment (kxhitiz) [reply _patterns/running-data-migration.md:40]: Same as [this](https://github.com/BiggerPockets/patterns/pull/182#discussion_r1419879532) comment

**Pull request merged**